x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java

```diff
@@ -177,13 +186,64 @@ public void getEngine(String resourceName, ActionListener<Engine> listener) {
         }));
     }
 
+    private static String getEngineAliasName(Engine engine) {
+        return ENGINE_ALIAS_PREFIX + engine.name();
+    }
+
     /**
      * Creates or updates the {@link Engine} in the underlying index.
      *
      * @param engine The engine object.
      * @param listener The action listener to invoke on response/failure.
      */
     public void putEngine(Engine engine, ActionListener<IndexResponse> listener) {
+        validateIndices(engine);
+
+        createOrUpdateAlias(engine, listener.delegateFailure((l, aliasResponse) -> { updateEngine(engine, l); }));
+
+    }
+
+    private void createOrUpdateAlias(Engine engine, ActionListener<AcknowledgedResponse> listener) {
+
+        final Metadata metadata = clusterService.state().metadata();
+        final String engineAliasName = getEngineAliasName(engine);
+
+        IndicesAliasesRequestBuilder requestBuilder = null;
+        if (metadata.hasAlias(engineAliasName)) {
+            Set<String> currentAliases = metadata.aliasedIndices(engineAliasName)
+                .stream()
+                .map(index -> index.getName())
+                .collect(Collectors.toSet());
+            Set<String> targetAliases = Set.of(engine.indices());
+
+            requestBuilder = updateAliasIndices(currentAliases, targetAliases, engineAliasName);
+
+        } else {
+            requestBuilder = clientWithOrigin.admin().indices().prepareAliases().addAlias(engine.indices(), engineAliasName);
+        }
+
+        requestBuilder.execute(listener);
+    }
+
+    private IndicesAliasesRequestBuilder updateAliasIndices(Set<String> currentAliases, Set<String> targetAliases, String engineAliasName) {
+
+        Set<String> deleteIndices = new HashSet<>(currentAliases);
+        deleteIndices.removeAll(targetAliases);
+
+        IndicesAliasesRequestBuilder aliasesRequestBuilder = clientWithOrigin.admin().indices().prepareAliases();
+
+        // Always re-add aliases, as an index could have been removed manually and it must be restored
```

Comment:
I apologize, I made that comment and then deleted it 🤦 I don't know if this is needed since you're pulling from the cluster metadata rather than the engine. Still better safe than sorry I suppose.

Reply:
I think it doesn't hurt - updating an engine shouldn't be that frequent, we can risk doing an additional update alias even if not needed.